pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go

```diff
@@ -274,50 +277,72 @@ func (n *klusterletController) sync(ctx context.Context, controllerContext facto
 		}
 	}
 
+	if klusterlet.DeletionTimestamp.IsZero() {
+		if !hasFinalezer(klusterlet, klusterletFinalizer) {
+			return n.addFinalezer(ctx, klusterlet, klusterletFinalizer)
+		}
+	}
+
 	// TODO: remove this when detached mode is not used in klusterlet
 	if config.InstallMode == operatorapiv1.InstallModeDetached {
 		config.InstallMode = operatorapiv1.InstallModeHosted
 	}
 
 	if config.InstallMode == operatorapiv1.InstallModeHosted {
+		var newStatus *operatorapiv1.KlusterletStatus
+		var updateErr error
 		managedClusterClients, err = n.buildManagedClusterClientsHostedMode(ctx, n.kubeClient, config.AgentNamespace, config.ExternalManagedKubeConfigSecret)
 		if err != nil {
-			_, _, _ = helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName, helpers.UpdateKlusterletConditionFn(metav1.Condition{
+			newStatus, _, updateErr = helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName, helpers.UpdateKlusterletConditionFn(metav1.Condition{
 				Type: klusterletReadyToApply, Status: metav1.ConditionFalse, Reason: "KlusterletPrepareFailed",
 				Message: fmt.Sprintf("Failed to build managed cluster clients: %v", err),
 			}))
-			return err
 		} else {
-			_, _, _ = helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName, helpers.UpdateKlusterletConditionFn(metav1.Condition{
+			newStatus, _, updateErr = helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName, helpers.UpdateKlusterletConditionFn(metav1.Condition{
 				Type: klusterletReadyToApply, Status: metav1.ConditionTrue, Reason: "KlusterletPrepared",
 				Message: "Klusterlet is ready to apply",
 			}))
 		}
+		if updateErr != nil {
+			return updateErr
+		}
+		klusterlet.Status = *newStatus
 	}
 
-	// Update finalizer after the clients created, otherwise, for Hosted mode if users did not provide the
-	// external managed cluster kubeconfig, we can not delete the finalizer
-	if klusterlet.DeletionTimestamp.IsZero() {
-		hasFinalizer := false
-		for i := range klusterlet.Finalizers {
-			if klusterlet.Finalizers[i] == klusterletFinalizer {
-				hasFinalizer = true
-				break
+	// Klusterlet is deleting, we remove its related resources on managed and management cluster
+	if !klusterlet.DeletionTimestamp.IsZero() {
+		var removeManagedClusterResources = true
+
+		if config.InstallMode == operatorapiv1.InstallModeHosted {
+			if hasFinalezer(klusterlet, klusterletHostedFinalizer) {
+				if !meta.IsStatusConditionTrue(klusterlet.Status.Conditions, klusterletReadyToApply) {
+					// wait for the external managed kubeconfig to exist to clean up resources on the manged cluster
+					return nil
```

Comment:
what will trigger this to reconcile, secret?

Reply:
yes